### PR TITLE
Test forceLocation key to make sure it exists.

### PR DESCRIPTION
**Pull request by tedsharpe**
Also add mock objectstore to test this.

### CODE REVIEW

File: boss-ws/src/main/java/org/genomebridge/boss/http/objectstore/ObjectStoreConfiguration.java [hunk 1]
```diff
@@ -20,6 +22,12 @@
...
 
+    public enum ObjectStoreType {
+        S3,  // Amazon S3 or ECS
+        GCS, // Google Cloud Storage
+        FCS  // Faux Cloud Storage
+    }
+
     @NotNull
-    public String type; // currently either S3 or GCS
+    public ObjectStoreType type; // currently either S3 or GCS
```
Comment (jmthibault79) [inline boss-ws/src/main/java/org/genomebridge/boss/http/objectstore/ObjectStoreConfiguration.java]: Comment is no longer accurate

File: boss-ws/src/main/java/org/genomebridge/boss/http/BossApplication.java [hunk 2]
```diff
@@ -95,11 +99,12 @@ public static BossAPI getAPI() {
...
-            return new S3ObjectStore(config);
-        if ( "GCS".equals(config.type) )
-            return new GCSObjectStore(config);
-        throw new IllegalStateException("ObjectStore configuration has unrecognized type: "+config.type);
+        switch ( config.type ) {
+        case S3: return new S3ObjectStore(config);
+        case GCS: return new GCSObjectStore(config);
+        case FCS: return new FCSObjectStore(config);
+        }
+        throw new IllegalArgumentException("No handler for ObjectStoreType "+config.type.toString());
```
Comment (jmthibault79) [inline boss-ws/src/main/java/org/genomebridge/boss/http/BossApplication.java:107]: Can this ever be reached, since it's not a vaild ObjectStoreType enum value?  My guess is that the failure would occur at the time of enum value creation instead.

Comment (tedsharpe) [inline boss-ws/src/main/java/org/genomebridge/boss/http/BossApplication.java:107]: It can't be reached right now, but if someone added a value to the enum and forgot to fix this switch it could be.  Just trying to be defensive.

File: boss-ws/src/main/java/org/genomebridge/boss/http/objectstore/S3ObjectStore.java [hunk 3]
```diff
@@ -18,16 +18,10 @@
...
-            AWSCredentials creds = new BasicAWSCredentials(config.username, config.password);
-            client = new AmazonS3Client(creds);
-        } else {
-            client = new AmazonS3Client();
-        }
-        if ( config.endpoint != null ) {
-            client.setEndpoint(config.endpoint);
-        }
-        if ( config.pathStyleAccess != null && config.pathStyleAccess ) {
+        AWSCredentials creds = new BasicAWSCredentials(config.username, config.password);
```
Comment (jmthibault79) [inline boss-ws/src/main/java/org/genomebridge/boss/http/objectstore/S3ObjectStore.java:21]: Why these changes?

Comment (tedsharpe) [inline boss-ws/src/main/java/org/genomebridge/boss/http/objectstore/S3ObjectStore.java:21]: We now have @NotNull marks on all ObjectStoreConfiguration properties, so the null testing is superfluous.  This old code (no username or password) was to support using an Amazon S3 bucket from "secret" credentials in a local .aws file.  Now that we have a mock object store people don't need any of this extra stuff to run unit tests, and that's better because we don't have to explain anything about how to set up to run unit tests -- they just run.

File: boss-ws/src/main/java/org/genomebridge/boss/http/service/DatabaseBossAPI.java [hunk 4]
```diff
@@ -73,12 +73,16 @@ public ErrorDesc insertObject(ObjectDesc rec, String userName) {
...
+        // or if the user is forcing the location to a known, pre-existing key,
         // otherwise generate a new (fresh) location.
         String loc = rec.directoryPath;
-        if ( !rec.storagePlatform.equals(StoragePlatform.OPAQUEURI.getValue()) &&
-                !Boolean.TRUE.equals(rec.forceLocation) )
-            loc = createLocation(rec);
+        if ( !rec.storagePlatform.equals(StoragePlatform.OPAQUEURI.getValue()) ) {
+            if ( !Boolean.TRUE.equals(rec.forceLocation) )
+                loc = createLocation(rec);
+            else if ( !getObjectStore(rec.storagePlatform).exists(loc) )
```
Comment (jmthibault79) [inline boss-ws/src/main/java/org/genomebridge/boss/http/service/DatabaseBossAPI.java:83]: `getObjectStore()` can return `null`, so `getObjectStore().exists()` might throw NPE

File: boss-ws/src/test/java/org/genomebridge/boss/http/AllObjectsAcceptanceTest.java [hunk 5]
```diff
@@ -192,22 +199,64 @@ public void testFindByName() {
     public void testForceLocation() {
         Client client = new Client();
 
+        // create an object
         ObjectDesc obj = new ObjectDesc();
         obj.ownerId = "fred";
         obj.objectName = "john";
-        String[] actors = new String[1];
```
Comment (jmthibault79) [inline boss-ws/src/test/java/org/genomebridge/boss/http/AllObjectsAcceptanceTest.java:198]: Thanks for removing something called "actors" :)

File: boss-ws/src/main/java/org/genomebridge/boss/http/BossApplication.java (hunk 2 above)
Comment (jmthibault79) [inline boss-ws/src/main/java/org/genomebridge/boss/http/BossApplication.java:107]: OK

File: boss-ws/src/main/java/org/genomebridge/boss/http/objectstore/S3ObjectStore.java (hunk 3 above)
Comment (jmthibault79) [inline boss-ws/src/main/java/org/genomebridge/boss/http/objectstore/S3ObjectStore.java:21]: Got it, thanks.

**Comment by jmthibault79** [comment]
Looks good.  The FCS is great!  A few remaining comments.
